doc/internals.rst

```diff
@@ -231,3 +231,448 @@ re-open it directly with Zarr:
+
+Some of the transformations can be common to more backends, so before
+implementing a new decoder, be sure Xarray does not already implement that one.
+
+The backends can reuse Xarray’s decoders, either instantiating the coders
+and using the method ``decode`` directly or using the higher-level function
+:py:func:`~xarray.conventions.decode_cf_variables` that groups Xarray decoders.
+
+In some cases, the transformation to apply strongly depends on the on-disk
+data format. Therefore, you may need to implement your own decoder.
+
+An example of such a case is when you have to deal with the time format of a
+grib file. grib format is very different from the NetCDF one: in grib, the
+time is stored in two attributes dataDate and dataTime as strings. Therefore,
+it is not possible to reuse the Xarray time decoder, and implementing a new
+one is mandatory.
+
+Decoders can be activated or deactivated using the boolean keywords of
+Xarray :py:meth:`~xarray.open_dataset` signature: ``mask_and_scale``,
+``decode_times``, ``decode_timedelta``, ``use_cftime``,
+``concat_characters``, ``decode_coords``.
+Such keywords are passed to the backend only if the User sets a value
+different from ``None``.  Note that the backend does not necessarily have to
+implement all the decoders, but it shall declare in its ``open_dataset``
+interface only the boolean keywords related to the supported decoders.
+
+.. _RST backend_registration:
+
+How to register a backend
++++++++++++++++++++++++++++
+
+Define a new entrypoint in your ``setup.py`` (or ``setup.cfg``) with:
+
+- group: ``xarray.backend``
+- name: the name to be passed to :py:meth:`~xarray.open_dataset`  as ``engine``
+- object reference: the reference of the class that you have implemented.
+
+You can declare the entrypoint in ``setup.py`` using the following syntax:
+
+.. code-block::
+
+    setuptools.setup(
+        entry_points={
+            "xarray.backends": [
+                "engine_name=your_package.your_module:your_backendentrypoint"
+            ],
+        },
+    )
+
+in ``setup.cfg``:
+
+.. code-block:: cfg
+
+    [options.entry_points]
+    xarray.backends =
+        engine_name = your_package.your_module:your_backendentrypoint
+
+
+See https://packaging.python.org/specifications/entry-points/#data-model
+for more information
+
+.. _RST lazy_loading:
+
+How to support Lazy Loading
++++++++++++++++++++++++++++
+If you want to make your backend effective with big datasets, then you should
+support lazy loading.
+Basically, you shall replace the :py:class:`numpy.ndarray` inside the
+variables with a custom class that supports lazy loading indexing.
+See the example below:
+
+.. code-block:: python
+
+    backend_array = MyBackendArray()
+    data = indexing.LazilyOuterIndexedArray(backend_array)
+    var = xr.Variable(dims, data, attrs=attrs, encoding=encoding)
+
+Where:
+
+- :py:class:`~xarray.core.indexing.LazilyOuterIndexedArray` is a class
+  provided by Xarray that manages the lazy loading.
```

Comment:
To help make things easier to understand, could we rename this class from `LazilyOuterIndexedArray` to the more generic name `LazilyIndexedArray`? (It currently works in all cases, not just outer indexing)

Reply:
done